Build the standard library crates when building the runtimes

koute opened the pull request:
Our executor currently only supports the WASM MVP feature set, however nowadays when compiling WASM the Rust compiler has more features enabled by default.

We do set the `-C target-cpu=mvp` flag to make sure that *our* code gets compiled in a way that is compatible with our executor, however this doesn't affect Rust's standard library crates (`std`, `core` and `alloc`) which are by default precompiled and still can make use of these extra features.

So in this PR we force the compiler to also compile the standard library crates for us to make sure that they also only use the MVP features.

I've added the `WASM_BUILD_STD` environment variable which can be used to disable this behavior if set to `0`.

Unfortunately this *will* slow down the compile times when building runtimes, but there isn't much that we can do about that.

Fixes https://github.com/paritytech/polkadot-sdk/issues/1755

liamaharon reviewed (commented):
Dumb question but isn't this only a problem building with the Debug profile? Or is it likely to leak into Release profile builds too?

bkchr reviewed (commented):
Generally looking okay, but could you also add some check if `rust-src` is installed and if not prints some instructions on how to install it. (Like it is done in prerequisites for the wasm target)

koute commented:
> Dumb question but isn't this only a problem building with the Debug profile? Or is it likely to leak into Release profile builds too?

Currently yes. But only by accident; there's no guarantee that the release profile won't also be hit by this issue. So I've made this the official way to build a runtime.

You can always use the `WASM_BUILD_STD` environment variable if you don't want this behavior.

> but could you also add some check if `rust-src` is installed and if not prints some instructions on how to install it. (Like it is done in prerequisites for the wasm target)

Done!

I've also simplified and improved the way we're doing the check right now. Funnily enough we have a whole song and dance there to extract the rustc version and there's even the following comment there:

> The best solution would be if there is a `cargo rustc --version` command, which sadly doesn't exists.

... which actually isn't correct, because this command does exist, so I refactored the code to use it. :P

I've also improved the error message when the `wasm32-unknown-unknown` is missing.

Previous error message:

```
error: failed to run custom build command for `node-template-runtime v4.0.0-dev (/local/cpp/polkadot-sdk/substrate/bin/node-template/runtime)`
note: To improve backtraces for build dependencies, set the CARGO_PROFILE_RELEASE_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug information generation.

Caused by:
  process didn't exit successfully: `/local/cpp/polkadot-sdk/target/release/build/node-template-runtime-dee511cb28c68720/build-script-build` (exit status: 1)
  --- stderr
  Rust WASM toolchain not installed, please install it!

  Further error information:
  ------------------------------------------------------------
     Compiling wasm-test v1.0.0 (/tmp/.tmp5lB5sY)
  error[E0463]: can't find crate for `std`
    |
    = note: the `wasm32-unknown-unknown` target may not be installed
    = help: consider downloading the target with `rustup target add wasm32-unknown-unknown`

  error: requires `sized` lang_item

  For more information about this error, try `rustc --explain E0463`.
  error: cannot find macro `println` in this scope
   --> src/main.rs:3:5
    |
  3 |                 println!("{}", env!("RUSTC_VERSION"));
    |                 ^^^^^^^

  error: could not compile `wasm-test` (lib) due to 2 previous errors
  warning: build failed, waiting for other jobs to finish...
  error: could not compile `wasm-test` (bin "wasm-test") due to 3 previous errors
  ------------------------------------------------------------

warning: build failed, waiting for other jobs to finish...
```

New error message:

```
error: failed to run custom build command for `node-template-runtime v4.0.0-dev (/local/cpp/polkadot-sdk/substrate/bin/node-template/runtime)`
note: To improve backtraces for build dependencies, set the CARGO_PROFILE_RELEASE_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug information generation.

Caused by:
  process didn't exit successfully: `/local/cpp/polkadot-sdk/target/release/build/node-template-runtime-dee511cb28c68720/build-script-build` (exit status: 1)
  --- stderr
  Cannot compile the WASM runtime: the `wasm32-unknown-unknown` target is not installed!
  You can install it with `rustup target add wasm32-unknown-unknown` if you're using `rustup`.
```

dzmitry-lahoda commented:
> Unfortunately this will slow down the compile times when building runtimes, but there isn't much that we can do about that.

did you though of wasm rewrite? most extension can be boiled down to simpler set? 

for example https://github.com/CosmWasm/cosmwasm project provides tool to comple floats to simulation.

bkchr commented:
> did you though of wasm rewrite? most extension can be boiled down to simpler set?

This slows down the compilation of the runtime.

bkchr commented on substrate/utils/wasm-builder/src/lib.rs:
```diff
@@ -282,6 +288,12 @@ impl CargoCommand {
 
+	/// Returns whether this version of the toolchain supports nightly features.
+	fn supports_nightly_features(&self) -> bool {
+		self.version.map(|version| version.is_nightly).unwrap_or(false) ||
```
```suggestion
		self.version.map_or(false |v| v.is_nightly) ||
```

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4491665

koute commented on substrate/utils/wasm-builder/src/lib.rs:
```diff
@@ -15,99 +15,7 @@
-//! installed as well. For example if installing the rust from 20.02.2020 using `rustup
-//! install nightly-2020-02-20`, the wasm target needs to be installed as well `rustup target add
-//! wasm32-unknown-unknown --toolchain nightly-2020-02-20`.
+#![doc = include_str!("../README.md")]
```
Hmm.... okay, so, do we want to get rid of the README completely? I can easily make another PR. :P

bkchr replied:
Yeah :P

koute replied:
Okay, no problem; I'll make a PR tomorrow. :P
